File: day9/part2.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

//#define TEST
#ifdef TEST
#define PLAYER_COUNT 30
#define LAST_MARBLE 5807
#else
#define PLAYER_COUNT 423
#define LAST_MARBLE 7194400
#endif

int main() {
    int* circle = (int*)malloc(sizeof(int) * (LAST_MARBLE + 3));
    if (circle == NULL)
        puts("oh no");
    int players[PLAYER_COUNT];
    for (int i = 0; i < PLAYER_COUNT; i++)
        players[i] = 0;
    int curMarbleI = 0;
    int count = 1;
    int curPlayer = 0;
    circle[0] = 0;

    for (int curMarble = 1; curMarble <= LAST_MARBLE; curMarble++) {
        if (curMarble % 23 > 0) {
            curMarbleI += 2;
            if (curMarbleI > count)
                curMarbleI -= count;
            memmove(circle + curMarbleI + 1, circle + curMarbleI, sizeof(int) * (count - curMarbleI));
            circle[curMarbleI] = curMarble;
            count++;
        }
        else {
            curMarbleI -= 7;
            if (curMarbleI < 0)
                curMarbleI += count;
            int removed = circle[curMarbleI];
            memmove(circle + curMarbleI, circle + curMarbleI + 1, sizeof(int) * (count - curMarbleI));
            count--;
            players[curPlayer] += curMarble + removed;
        }
        curPlayer = (curPlayer + 1) % PLAYER_COUNT;
    }

    int best = 0;
    for (int i = 0; i < PLAYER_COUNT; i++) {
        if (players[i] > players[best])
            best = i;
    }
    printf("%d\n", players[best]);
    return 0;
}

File: day9/part2.js

function getHighscore(playerCount, lastMarble) {
    const players = new Array(playerCount).fill(0);
    let circle = [ 0 ];
    let currentMarbleI = 0;
    let currentPlayer = 0;
    for (let curMarble = 1; curMarble <= lastMarble; curMarble++) {
        if (curMarble % 23 > 0) {
            currentMarbleI = (currentMarbleI + 2);
            if (currentMarbleI > circle.length)
                currentMarbleI -= circle.length;
            circle = circle.slice(0, currentMarbleI).concat(curMarble).concat(circle.slice(currentMarbleI));
        }
        else {
            currentMarbleI -= 7;
            if (currentMarbleI < 0)
                currentMarbleI += circle.length;
            players[currentPlayer] += curMarble + circle.splice(currentMarbleI, 1)[0];
        }
        currentPlayer = (currentPlayer + 1) % players.length;
    }

    let best = 0;
    for (let i = 0; i < players.length; i++) {
        if (players[i] > players[best])
            best = i;
    }
    return players[best];
}

const testInput = [
    { players: 10, lastMarble: 1618, expected: 8317 },
    { players: 13, lastMarble: 7999, expected: 146373 },
    { players: 17, lastMarble: 1104, expected: 2764 },
    { players: 21, lastMarble: 6111, expected: 54718 },
    { players: 30, lastMarble: 5807, expected: 37305 }
];
const realInput = { players: 423, lastMarble: 7194400 };

for (const test of testInput) {
    const actual = getHighscore(test.players, test.lastMarble);
    console.log(`Expected: ${test.expected}\tActual: ${actual}\t${actual===test.expected ? "SUCCESS" : "FAILED"}`);
}
console.log("Solution: ", getHighscore(realInput.players, realInput.lastMarble));
